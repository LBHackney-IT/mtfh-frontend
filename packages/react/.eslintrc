{
  "extends": "@lbh-mtfh/eslint-config",
  "parserOptions": {
    "project": "tsconfig.json"
  },
  "rules": {
    "prefer-arrow-callback": "off",
    "import/no-extraneous-dependencies": [
      "error",
      {
        "devDependencies": ["**/*/rollup.config.js"]
      }
    ],
    "jsx-a11y/anchor-is-valid": [
      "error",
      {
        "components": ["Link"],
        "aspects": ["invalidHref"]
      }
    ]
  }
}
