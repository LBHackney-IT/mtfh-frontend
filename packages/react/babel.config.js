const babel = require("../../babel.config")

module.exports = babel
