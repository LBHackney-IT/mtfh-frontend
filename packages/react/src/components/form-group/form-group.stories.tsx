import React from "react";

import { Input } from "../input";
import { FormGroup } from "./form-group";

export default {
  title: "Form Group",
  component: FormGroup,
};

export const Base = () => {
  return (
    <FormGroup id="name" label="First Name">
      <Input />
    </FormGroup>
  );
};

export const FormGroupWithHint = () => {
  return (
    <FormGroup
      id="name"
      label="First Name"
      hint="You can find this name on your passport"
    >
      <Input />
    </FormGroup>
  );
};

export const FormGroupWithRequired = () => {
  return (
    <FormGroup id="name" label="First Name" required>
      <Input />
    </FormGroup>
  );
};

export const FormGroupWithError = () => {
  return (
    <FormGroup
      id="name"
      label="First Name"
      error="First Name is required"
      required
    >
      <Input />
    </FormGroup>
  );
};
