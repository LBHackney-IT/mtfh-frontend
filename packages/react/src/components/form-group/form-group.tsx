import React, {
  Children,
  ComponentPropsWithoutRef,
  ReactElement,
  cloneElement,
  forwardRef,
  useMemo,
} from "react";
import classNames from "classnames";

import { widthOverrides } from "../../utils";
import { ErrorMessage } from "../error-message";
import { Hint } from "../hint";
import { Label } from "../label";
import styles from "./styles.module.scss";

export interface FormGroupProps extends ComponentPropsWithoutRef<"div"> {
  id: string;
  label?: string;
  name?: string;
  hint?: string;
  error?: string | false;
  required?: boolean;
  children: ReactElement;
  override?: number;
}

export const FormGroup = forwardRef<HTMLDivElement, FormGroupProps>(
  function FormGroup(
    {
      id,
      name,
      label,
      hint,
      error,
      required,
      children,
      className,
      override,
      ...props
    },
    ref
  ) {
    const formGroupClasses = classNames(
      styles.govukFormGroup,
      {
        [styles.govukFormGroupError]: !!error,
      },
      styles.lbhFormGroup,
      widthOverrides(override),
      className
    );

    const describedBy = useMemo(() => {
      const ids: string[] = [];
      if (hint) {
        ids.push(`${id}-hint`);
      }
      if (error) {
        ids.push(`${id}-error`);
      }
      return ids.join(" ");
    }, [id, hint, error]);

    return (
      <div ref={ref} id={id} className={formGroupClasses} {...props}>
        {!!label && (
          <Label htmlFor={`${id}-field`}>
            {label}
            {required ? <sup aria-hidden="true">*</sup> : ""}
          </Label>
        )}
        {!!hint && <Hint id={`${id}-hint`}>{hint}</Hint>}
        {!!error && <ErrorMessage id={`${id}-error`}>{error}</ErrorMessage>}
        {!!children &&
          Children.only(
            cloneElement(children, {
              id: `${id}-field`,
              name,
              required,
              error: !!error,
              "aria-describedby": describedBy || undefined,
              ...children.props,
            })
          )}
      </div>
    );
  }
);
