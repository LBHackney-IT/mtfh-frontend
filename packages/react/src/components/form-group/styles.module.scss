@import "~lbh-frontend/lbh/base";
@import "~govuk-frontend/govuk/objects/form-group";
@import "~lbh-frontend/lbh/objects/form-group";

.lbh-form-group .govuk-label sup {
  color: lbh-colour("lbh-error");
}
