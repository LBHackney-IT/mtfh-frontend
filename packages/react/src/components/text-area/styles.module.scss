@import "~lbh-frontend/lbh/base";
@import "~lbh-frontend/lbh/components/lbh-textarea/textarea";
@import "~lbh-frontend/lbh/components/lbh-character-count/character-count";

label + .govuk-character-count {
  margin-top: 0;
}
