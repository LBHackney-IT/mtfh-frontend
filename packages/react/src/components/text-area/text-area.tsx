import React, {
  ChangeEvent,
  ComponentPropsWithoutRef,
  forwardRef,
  useCallback,
  useMemo,
  useState,
} from "react";
import classNames from "classnames";

import { pluralize, widthOverrides } from "../../utils";
import { ErrorMessage } from "../error-message";
import { Hint } from "../hint";

import styles from "./styles.module.scss";

export interface TextAreaProps extends ComponentPropsWithoutRef<"textarea"> {
  maxLength?: number;
  error?: boolean;
  override?: number;
}

const getLengthOfValue = (
  initialValue: string | number | readonly string[] | undefined,
) => {
  if (typeof initialValue === "string") {
    return initialValue.length;
  }
  if (Array.isArray(initialValue)) {
    return initialValue.join(",").length;
  }
  return String(initialValue || "").length;
};

export const TextArea = forwardRef<HTMLTextAreaElement, TextAreaProps>(function TextArea(
  { maxLength, error, className, onChange, override, ...props },
  ref,
) {
  const { value, defaultValue } = props;
  const isControlled = value !== undefined;
  const initialValue = value || defaultValue;

  const [characterCount, setCharacterCount] = useState(getLengthOfValue(initialValue));

  const onChangeHandler = useCallback(
    (event: ChangeEvent<HTMLTextAreaElement>) => {
      if (
        event?.currentTarget?.value !== undefined &&
        !isControlled &&
        maxLength !== undefined
      ) {
        setCharacterCount(String(event.currentTarget.value).length);
      }

      if (typeof onChange === "function") {
        onChange(event);
      }
    },
    [onChange, maxLength, isControlled],
  );

  const exceedingValue = useMemo(
    () =>
      maxLength !== undefined &&
      maxLength - (isControlled ? getLengthOfValue(value) : characterCount),
    [maxLength, characterCount, value, isControlled],
  );

  const textareaClasses = classNames(
    styles.govukTextarea,
    styles.lbhTextarea,
    { [styles.govukTextareaError]: error },
    styles.govukCharacterCount,
    styles.lbhCharacterCount,
    widthOverrides(override),
    className,
  );

  const messageClasses = classNames(
    styles.govukCharacterCountMessage,
    widthOverrides(override),
  );

  const component = (
    <>
      <textarea
        ref={ref}
        className={textareaClasses}
        onChange={onChangeHandler}
        {...props}
      />
      {maxLength !== undefined && exceedingValue !== false && (
        <>
          {exceedingValue >= 0 ? (
            <Hint className={messageClasses} aria-live="polite">
              You have {exceedingValue} {pluralize("character", exceedingValue)} remaining
            </Hint>
          ) : (
            <ErrorMessage className={messageClasses} aria-live="polite">
              You have {Math.abs(exceedingValue)} {pluralize("character", exceedingValue)}{" "}
              too many
            </ErrorMessage>
          )}
        </>
      )}
    </>
  );

  return maxLength !== undefined ? (
    <div className={styles.govukCharacterCount}>{component}</div>
  ) : (
    component
  );
});
