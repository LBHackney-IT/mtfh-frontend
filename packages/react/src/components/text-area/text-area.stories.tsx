import React from "react";

import { FormGroup } from "../form-group";
import { TextArea } from "./text-area";

export default {
  title: "Textarea",
  component: TextArea,
};

export const Base = () => {
  return <TextArea rows={5} />;
};

export const TextareaWithCharacterCount = () => {
  return (
    <FormGroup id="count" label="Comment">
      <TextArea maxLength={200} />
    </FormGroup>
  );
};
