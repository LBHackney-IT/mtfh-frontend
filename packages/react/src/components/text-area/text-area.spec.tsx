import React, { ChangeEvent, useState } from "react";
import { fireEvent, render } from "@testing-library/react";
import { testA11y } from "@lbh-mtfh/test-utils";

import { FormGroup } from "../form-group";
import { TextArea } from "./text-area";

test("it passes a11y", async () => {
  const { container } = render(
    <FormGroup id="test" name="test" label="TextArea Test">
      <TextArea />
    </FormGroup>
  );
  expect(container).toMatchSnapshot();
  await testA11y(container);
});

test("it renders correctly", () => {
  const { container } = render(<TextArea />);
  expect(container).toMatchSnapshot();
});

test("it renders with character count if maxLength is set", () => {
  const { getByText } = render(<TextArea maxLength={500} />);
  expect(getByText("You have 500 characters remaining")).toBeInTheDocument();
});

test("it renders with character count with correct plural", () => {
  const { getByText, getByPlaceholderText } = render(
    <TextArea maxLength={2} placeholder="TextArea" />
  );
  const textArea = getByPlaceholderText("TextArea");

  expect(getByText("You have 2 characters remaining")).toBeInTheDocument();
  fireEvent.change(textArea, { target: { value: "A" } });
  expect(getByText("You have 1 character remaining")).toBeInTheDocument();
  fireEvent.change(textArea, { target: { value: "AB" } });
  expect(getByText("You have 0 characters remaining")).toBeInTheDocument();
  fireEvent.change(textArea, { target: { value: "ABC" } });
  expect(getByText("You have 1 character too many")).toBeInTheDocument();
  fireEvent.change(textArea, { target: { value: "ABCD" } });
  expect(getByText("You have 2 characters too many")).toBeInTheDocument();
});

test("it decreases the character count when text area value increases", () => {
  const { getByText, getByPlaceholderText } = render(
    <TextArea maxLength={500} placeholder="TextArea" />
  );
  const textArea = getByPlaceholderText("TextArea");
  fireEvent.change(textArea, { target: { value: "Some text" } });
  expect(getByText("You have 491 characters remaining")).toBeInTheDocument();
});

test("it renders with character count if component is controlled", () => {
  const { getByText } = render(<TextArea maxLength={500} value="Some text" />);
  expect(getByText("You have 491 characters remaining")).toBeInTheDocument();
});

test("it renders with character count with array value", () => {
  const { getByText } = render(
    <TextArea maxLength={500} defaultValue={["Some", "text"]} />
  );
  expect(getByText("You have 491 characters remaining")).toBeInTheDocument();
});

test("it renders with character count with number value", () => {
  const { getByText } = render(<TextArea maxLength={5} defaultValue={1234} />);
  expect(getByText("You have 1 character remaining")).toBeInTheDocument();
});

test("it renders with character count with undefined value", () => {
  const { getByText } = render(
    <TextArea maxLength={5} defaultValue={undefined} />
  );
  expect(getByText("You have 5 characters remaining")).toBeInTheDocument();
});

test("it supports an onChange handler", () => {
  const onChange = jest.fn<void, ChangeEvent<HTMLTextAreaElement>[]>();
  const { getByPlaceholderText } = render(
    <TextArea placeholder="TextArea" onChange={onChange} />
  );
  const textArea = getByPlaceholderText("TextArea");
  fireEvent.change(textArea, { target: { value: "Some text" } });
  expect(onChange).toHaveBeenCalledTimes(1);
  expect(onChange.mock.calls[0][0].target.value).toContain("Some text");
});

test("it supports an onChange handler with maxLength", () => {
  const onChange = jest.fn<void, ChangeEvent<HTMLTextAreaElement>[]>();
  const { getByPlaceholderText } = render(
    <TextArea maxLength={500} placeholder="TextArea" onChange={onChange} />
  );
  const textArea = getByPlaceholderText("TextArea");
  fireEvent.change(textArea, { target: { value: "Some text" } });
  expect(onChange).toHaveBeenCalledTimes(1);
  expect(onChange.mock.calls[0][0].target.value).toContain("Some text");
});

test("it supports a controlled component", () => {
  const ControlledComponentTest = () => {
    const [value, setValue] = useState("");
    return (
      <>
        <TextArea
          maxLength={10}
          value={value}
          onChange={(event: ChangeEvent<HTMLTextAreaElement>) =>
            setValue(event.currentTarget.value)
          }
          placeholder="TextArea"
        />
        <button type="button" onClick={() => setValue("Hello")}>
          Test button
        </button>
      </>
    );
  };
  const { getByPlaceholderText, getByText } = render(
    <ControlledComponentTest />
  );
  const textArea = getByPlaceholderText("TextArea") as HTMLTextAreaElement;
  fireEvent.change(textArea, { target: { value: "Some text" } });
  expect(textArea.value).toBe("Some text");
  const button = getByText("Test button");
  fireEvent.click(button);
  expect(textArea.value).toBe("Hello");
  expect(getByText("You have 5 characters remaining")).toBeInTheDocument();
});
