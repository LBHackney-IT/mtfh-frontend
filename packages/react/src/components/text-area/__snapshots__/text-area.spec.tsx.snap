// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it passes a11y 1`] = `
<div>
  <div
    class="govukFormGroup lbhFormGroup"
    id="test"
  >
    <label
      class="govukLabel lbhLabel"
      for="test-field"
    >
      TextArea Test
      
    </label>
    <textarea
      class="govukTextarea lbhTextarea govukCharacterCount lbhCharacterCount"
      id="test-field"
      name="test"
    />
  </div>
</div>
`;

exports[`it renders correctly 1`] = `
<div>
  <textarea
    class="govukTextarea lbhTextarea govukCharacterCount lbhCharacterCount"
  />
</div>
`;
